Spill/restore FP/BP around instructions in which they are clobbered

weiguozhi opened the pull request:
This patch fixes https://github.com/llvm/llvm-project/issues/17204.

If a base pointer is used in a function, and it is clobbered by an instruction (typically an inline asm), current register allocator can't handle this situation, so BP becomes garbage after those instructions. It can also occur to FP in theory.

We can spill and reload FP/BP registers around those instructions. But normal spill/reload instructions also use FP/BP, so we can't spill them into normal spill slots, instead we spill them into the top of stack by using SP register.

weiguozhi committed:
Spill/restore FP/BP around instructions in which they are clobbered

This patch fixes https://github.com/llvm/llvm-project/issues/17204.

If a base pointer is used in a function, and it is clobbered by an
instruction (typically an inline asm), current register allocator can't
handle this situation, so BP becomes garbage after those instructions.
It can also occur to FP in theory.

We can spill and reload FP/BP registers around those instructions. But
normal spill/reload instructions also use FP/BP, so we can't spill them
into normal spill slots, instead we spill them into the top of stack by
using SP register.

aengelke commented:
This apparently doesn't catch all cases. Example: ghccc uses rbp as argument register, but rbp is used after the function call despite being clobbered before.

```
; RUN: llc -mtriple=x86_64-pc-linux -stackrealign -verify-machineinstrs < %s

declare ghccc i32 @external(i32)

define i32 @foo(i32 %0, i32 %1) {
    %x = call ghccc i32 @external(i32 %0, i32 %1)
    ret i32 %x
}
```
Output (w/o directives for brevity):
```
foo:                                    # @foo
# %bb.0:
        pushq   %rbp
        movq    %rsp, %rbp
        pushq   %r15
        pushq   %r14
        pushq   %r13
        pushq   %r12
        pushq   %rbx
        andq    $-16, %rsp
        subq    $16, %rsp
                                        # kill: def $edi killed $edi def $rdi
        movl    %esi, %ebp
        movq    %rdi, %r13
        callq   external@PLT
        leaq    -40(%rbp), %rsp
        popq    %rbx
        popq    %r12
        popq    %r13
        popq    %r14
        popq    %r15
        popq    %rbp
        retq
```

weiguozhi committed:
Spill/restore FP/BP around instructions in which they are clobbered

Add a new test case to demostrate save/restore of FP register.

weiguozhi commented:
Thanks for the test case. I added it to this patch.

This is exactly what I had in my mind, see the comments for function spillFrameBasePointer. But I failed to find a real test case. So I disabled checking fp by default. It can be enabled with -spill-clobbered-fp=true, and then I got

```
        pushq   %rbp
        movq    %rsp, %rbp
        pushq   %r15
        pushq   %r14
        pushq   %r13
        pushq   %r12
        pushq   %rbx
        andq    $-16, %rsp
        subq    $16, %rsp
        pushq   %rbp                   // save rbp
        pushq   %rax
        movl    %esi, %ebp
        movq    %rdi, %r13
        callq   external@PLT
        addq    $8, %rsp
        popq    %rbp                   // restore rbp
        leaq    -40(%rbp), %rsp
        popq    %rbx
        popq    %r12
        popq    %r13
        popq    %r14
        popq    %r15
        popq    %rbp
        retq
```

weiguozhi commented:
ping

aengelke commented:
Still, the CFI info looks wrong once the frame pointer is clobbered (updates to rsp *and* rbp should be accompanied by matching CFI directives) and I have concerns that this will break unwinding (not just asynchronous unwinding but also exceptions thrown from the called function). Could you test that unwinding through such a function call works as expected?

(fwiw, I'm not a maintainer and I'd personally like to see support for this use case, but I have concerns that this will break other things.)

rnk commented on llvm/include/llvm/CodeGen/TargetRegisterInfo.h:
```diff
@@ -963,6 +963,16 @@ class TargetRegisterInfo : public MCRegisterInfo {
     return false;
   }
 
+  /// Returns true if the target needs a base register.
```
I noticed that AArch64, PPC, ARM, and RISCV all use virtual base registers (requiresVirtualBaseRegisters returns true in most cases) instead of physical base registers. Should we look into that in addition to this change? Perhaps we can add an x86-specific cl::opt that uses a virtual base register, try that on some corpus of code, and see if that is a) correct and b) has acceptable performance.

Regardless, we still need support for spilling the FP around FP-clobbers, so I think we need your change, and we should go forward as is.

rnk commented on llvm/lib/CodeGen/PrologEpilogInserter.cpp:
```diff
@@ -1587,3 +1600,114 @@ void PEI::replaceFrameIndices(MachineBasicBlock *BB, MachineFunction &MF,
+/// register as base pointer and let RA allocate a physical register to it.
+/// 2. If there is no other instructions access stack with fp/bp from the
+/// inline asm to the epilog, we can skip the save and restore operations.
+void PEI::spillFrameBasePointer(MachineFunction &MF) {
```
I think the physical base pointer concept is essentially specific to x86. Is there a hook we can use to move this entire pass into X86FrameLowering? That would avoid the need to add as many hooks to TargetFrameLowering and TargetRegisterInfo.

weiguozhi replied:
There are also many targets don't implement requiresVirtualBaseRegisters, I don't know if they are also using physical base pointer. The frame pointer part should impact almost all targets.

rnk reviewed (commented):
Thanks for working on this! This issue is over a decade old.

weiguozhi commented:
> Still, the CFI info looks wrong once the frame pointer is clobbered (updates to rsp _and_ rbp should be accompanied by matching CFI directives) and I have concerns that this will break unwinding (not just asynchronous unwinding but also exceptions thrown from the called function). Could you test that unwinding through such a function call works as expected?
> 
I'm very ignorant in DWARF. After some study, I think we need the following additional CFI instructions

```
foo:                                    # @foo
        .cfi_startproc
# %bb.0:
        pushq   %rbp
        .cfi_def_cfa_offset 16
        .cfi_offset %rbp, -16
        movq    %rsp, %rbp
        .cfi_def_cfa_register %rbp
        pushq   %r15
        pushq   %r14
        pushq   %r13
        pushq   %r12
        pushq   %rbx
        andq    $-16, %rsp
        subq    $16, %rsp
        .cfi_offset %rbx, -56
        .cfi_offset %r12, -48
        .cfi_offset %r13, -40
        .cfi_offset %r14, -32
        .cfi_offset %r15, -24
                                        # kill: def $edi killed $edi def $rdi
        pushq   %rbp
        pushq   %rax
+        .cfi_def_cfa_expression    [%rsp + 8]   + 16              //  %rbp is stored in memory [%rsp + 8]
        movl    %esi, %ebp
        movq    %rdi, %r13
        callq   external@PLT
        addq    $8, %rsp
        popq    %rbp
+        .cfi_def_cfa   %rbp, 16                                   // %rbp is restored
        leaq    -40(%rbp), %rsp
        popq    %rbx
        popq    %r12
        popq    %r13
        popq    %r14
        popq    %r15
        popq    %rbp
        .cfi_def_cfa %rsp, 8
        retq
```
And the cfi_def_cfa_expression should be constructed similar to function X86FrameLowering::emitCalleeSavedFrameMoves. Is my understanding right?

> (fwiw, I'm not a maintainer and I'd personally like to see support for this use case, but I have concerns that this will break other things.)

Because the frame pointer is clobbered, so the unwinding information is already broken. The current patch doesn't make it worse.

rnk reviewed (commented):
I'm finding myself short on time, so maybe I can hand the review off to @zmodem , who has done various backend changes to switch lowering and conditional tail calls and so on.

The current behavior is pretty bad, so even if there are some edge cases that this solution doesn't handle such as inaccurate unwind info, this will be an improvement, and we should try to move forward.

weiguozhi committed:
Add unwinding information when FP is changed.

weiguozhi commented:
I added unwinding information when FP is saved/restored.

rnk commented:
Hans, can you please review this?

zmodem reviewed (commented):
Apologies for the delay. I'll do my best to review this.

Since this is currently only handling x86, should we keep it in the x86 target for now, and try to generalize later (maybe with some input from an aarch64 owner)?

weiguozhi commented:
> Since this is currently only handling x86, should we keep it in the x86 target for now, and try to generalize later (maybe with some input from an aarch64 owner)?

I will move it to x86 backend for this initial version. Then many of the callbacks can be removed.

efriedma-quic commented on llvm/lib/Target/X86/X86FrameLowering.cpp:
```diff
@@ -4178,3 +4178,158 @@ void X86FrameLowering::restoreWinEHStackPointersInParent(
+    BuildMI(*MBB, BeforeMI, DL, TII.get(TargetOpcode::CFI_INSTRUCTION))
+        .addCFIIndex(CFIIndex);
+
+    // Setup new CFA value with DW_CFA_def_cfa_expression:
```
If you're spilling the base pointer, you shouldn't need to touch the unwind: we normally restore the stack pointer from fp, anyway.

If you're spilling the frame pointer, you'd have to do something complicated to get correct unwind; probably represent it as a separate function (.seh_endproc/.seh_startproc/etc.).  You can't specify unwind opcodes outside the prolog.

weiguozhi committed:
Move the implementation to X86 backend.

weiguozhi commented:
I have moved the implementation to x86 backend. Please review it again.

david-xl commented:
Any more review comments for this patch? The underlying problem is blocking the preserve_none CC from being widely used.

rnk commented on llvm/lib/Target/X86/X86FrameLowering.cpp:
```diff
@@ -4231,3 +4231,325 @@ void X86FrameLowering::restoreWinEHStackPointersInParent(
+    FP = TRI->getFrameRegister(MF);
+    if (STI.isTarget64BitILP32())
+      FP = Register(getX86SubSuperRegister(FP, 64));
+    RC = TRI->getMinimalPhysRegClass(FP);
```
This register class calculation is the same between spill and restore FPBP.

I suggest you can restructure this to have one method, `saveAndRestoreFPBPUsingSP`, which takes `BeforeMI` and `AfterMI`, calculates this shared state once, and then calls `spill/restoreFPBPUsingSP`, and maybe just pass in the `FP/BP` Register variables. They are nullable, so you can check for `isValid` before pushing or popping.

`SPAdjust` will also be the same and only needs to be calculated once.

weiguozhi replied:
done.

rnk commented on llvm/lib/CodeGen/PrologEpilogInserter.cpp:
```diff
@@ -228,6 +228,8 @@ bool PEI::runOnMachineFunction(MachineFunction &MF) {
   FrameIndexVirtualScavenging = TRI->requiresFrameIndexScavenging(MF);
   ORE = &getAnalysis<MachineOptimizationRemarkEmitterPass>().getORE();
 
+  TFI->spillFPBP(MF);
```
The PEI process is very complicated, please add comments elaborating on what this does and why it comes before call frame instruction elimination.

weiguozhi replied:
done.

rnk commented on llvm/lib/Target/X86/X86FrameLowering.cpp:
```diff
@@ -4231,3 +4231,325 @@ void X86FrameLowering::restoreWinEHStackPointersInParent(
+
+// Invoke instruction has been lowered to normal function call. We try to figure
+// out if MI comes from Invoke.
+// Do we have any better method?
```
This feels like an incomplete solution. If we have an invoke that actually clobbers RBX and we need to spill around it, we ought to reliably fail with an error instead of silently doing nothing as we currently do. That's part of what I really want to get out of this. We can't solve all problems, but the compiler should emit an error if it gets into an impossible situation that it can't handle.

weiguozhi replied:
Good point. I will do detection work in a separate patch.

rnk replied:
I can live with that.

rnk commented on llvm/test/CodeGen/X86/clobber_frame_ptr.ll:
```diff
@@ -0,0 +1,52 @@
+; CHECK-NEXT:    popq %rbp
+; CHECK-NEXT:    .cfi_def_cfa %rsp, 8
+; CHECK-NEXT:    retq
+    %x = call ghccc i32 @external(i32 %0, i32 %1)
```
Is there any other calling convention that:
1. clobbers RBX
2. accepts stack arguments

I'd like to exercise the test case where:
* we need to pass stack arguments (num regparm args + 4)
* we have to spill *just* FP (RBP, not RBX)

I'm worried that the stores or pushes to set up the stack arguments will be disturbed by the RBP push and pops.

Regardless of whether the problem exists or not, please add a test case that exercises spilling around a function that clobbers FP/BP with stack arguments.

weiguozhi replied:
The only calling convention that clobbers RBP but no RBX is HiPE. I added a new test case for it. And this patch does correct work.

But the generated code is still wrong because of a different bug.
```
  %6:gr64 = MOV64rm %fixed-stack.1, 1, $noreg, 0, $noreg :: (load (s64) from %fixed-stack.1)
  %7:gr64 = MOV64rm %fixed-stack.0, 1, $noreg, 0, $noreg :: (load (s64) from %fixed-stack.0)
  ADJCALLSTACKDOWN64 16, 0, 0, implicit-def dead $rsp, implicit-def dead $eflags, implicit-def dead $ssp, implicit $rsp, implicit $ssp
  %8:gr64 = COPY $rsp 
  MOV64mr %8:gr64, 1, $noreg, 8, $noreg, killed %7:gr64 :: (store (s64) into stack + 8) 
  MOV64mr %8:gr64, 1, $noreg, 0, $noreg, killed %6:gr64 :: (store (s64) into stack)
  $r15 = COPY %0:gr64
  $rbp = COPY %1:gr64
  $rsi = COPY %2:gr64
  $rdx = COPY %3:gr64
  $rcx = COPY %4:gr64
  $r8 = COPY %5:gr64
  CALL64pcrel32 target-flags(x86-plt) @external2, <regmask>, ...
```
x86-cf-opt transforms it to
```
  ADJCALLSTACKDOWN64 16, 0, 16, implicit-def dead $rsp, implicit-def dead $eflags, implicit-def dead $ssp, implicit $rsp, implicit $ssp 
  $r15 = COPY %0:gr64
  $rbp = COPY %1:gr64
  $rsi = COPY %2:gr64
  $rdx = COPY %3:gr64
  $rcx = COPY %4:gr64
  $r8 = COPY %5:gr64
  PUSH64rmm %fixed-stack.0, 1, $noreg, 0, $noreg, implicit-def $rsp, implicit $rsp :: (load (s64) from %fixed-stack.0), (store (s64) into stack + 8)
  PUSH64rmm %fixed-stack.1, 1, $noreg, 0, $noreg, implicit-def $rsp, implicit $rsp :: (load (s64) from %fixed-stack.1), (store (s64) into stack)
  CALL64pcrel32 target-flags(x86-plt) @external2, <regmask>, ...
```
The access to %fixed-stack.0 needs rbp, but it is already assigned function argument.

rnk replied:
I see, that's a good find. :) Thanks for adding the tests, we don't need to fix this issue before landing.

rnk reviewed (commented):
Hans is out on summer break, but I went ahead and took a look

weiguozhi committed:
Refactor the code

Move the duplicated computation of physical register, register class and
stack adjustment to new function saveAndRestoreFPBPUsingSP.

weiguozhi commented on llvm/lib/Target/X86/X86FrameLowering.cpp:
```diff
@@ -4231,3 +4231,325 @@ void X86FrameLowering::restoreWinEHStackPointersInParent(
   }
 }
+
+static int computeSPAdjust4SpillFPBP(MachineFunction &MF,
```
done.

rnk commented on llvm/lib/CodeGen/CFIInstrInserter.cpp:
```diff
@@ -228,17 +232,23 @@ void CFIInstrInserter::calculateOutgoingCFAInfo(MBBCFAInfo &MBBInfo) {
+        // the same BB, so it will not impact outgoing CFA.
+        ++RememberState;
+        if (RememberState != 1)
+          report_fatal_error(
```
This is existing code, but I generally try to recommend `MCContext::reportError` over `report_fatal_error`. We do have actual non-fatal error handling facilities during CodeGen.

weiguozhi replied:
done.

rnk reviewed (approved):
Thanks, let's merge it!

weiguozhi committed:
Replace report_fatal_error with MCContext::reportError

weiguozhi committed:
Fix format

rnk reviewed (approved):
Thanks again!

weiguozhi committed:
Merge branch 'main' into carrot-base-ptr

aengelke commented:
This incurs a substantial compile-time regression at O0. http://llvm-compile-time-tracker.com/compare.php?from=ce2a3d9042c95630f12b790bf201c4daf8941afb&to=0d471b3f64d3116bd57c79d872f7384fff80daa5&stat=instructions:u

Can we come up with a no-overhead solution for this, e.g. check during call lowering whether there's a call that clobbers FP/BP and then selectively look at these calls instead of iterating over the entire IR? (IR iteration is expensive. And >99% of the code doesn't need this.) If this takes longer, we might want to temporarily revert this.

rnk commented:
I think what you suggest is feasible. I would try to skip this pass under the following conditions:
* does not use base pointer (BP, but may use FP)
* has no inline asm (we already track this somewhere)
* has no FP-clobbering function calls. This is a new flag we could add to X86MachineFunctionInfo and set in X86ISelLoweringCall.cpp somewhere.

The vast majority of functions do not do both realign the stack in the prologue and adjust SP (dynamic alloca) after the prologue, necessitating a base pointer. -O0 functions do tend to have FPs, so it's worth adding an extra flag to track FP clobbering so we can early-exit in those cases. Note that the non-O0 compile time are not affected, presumably they use `fomit-frame-pointer`.

weiguozhi commented:
@aengelke @rnk thanks for the good suggestion. I will implement it in a new patch.

Current implementation already checks if BP/FP is used in the function, and skip the pass if neither is used. So it seems there are still a lot of functions use FP with -O0.

bylaws commented:
Is this intended to impact __attribute__(__naked__)) functions? With msvc x86 this attr causes asm to be emitted exactly as specified and clang diverges here.

efriedma-quic commented:
No, that's a bug; please open an issue.
